### ISSUE TITLE
There should be an AbstractObject class

### ISSUE THREAD

**Issue by telephon** [supercollider/supercollider#3467]
We could extend sclang's flexibilty considerably if we'd added `AbstractObject`, a superclass for `Object`. This class would only support a minimal interface of methods (mainly for introspection). This would follow the idea of Ruby's [BasicObject](https://ruby-doc.org/core-2.0.0/BasicObject.html).

By default, all classes would still inherit from `Object`. By inheriting from `AbstractObject`, it would be possible to build alternative behaviours for objects which are not currently possible.

The most immediate practical impact would be that we could make Object prototyping safe (currently, pseudo-methods may be overridden later by a class extension of any class above `Event`).

I have written number of experimental classes with a similar approach, which is to be found here: https://github.com/telephon/Neutral

The current inheritance scheme:
`Object ––superclass–> nil`
`Meta_Object ––superclass–> Class`

The new inheritance scheme would be:
`Object ––superclass–> AbstractObject ––superclass–> nil`
`Meta_Object ––superclass–> Meta_AbstractObject ––superclass–> Class`

Side Note: a primitive for directly forwarding a method call to a different receiver unchanged would make the implementation very efficient, and could make it an alternative to method extensions.

(was: #1831)

**Comment by scztt** [comment]
FYI method forwarding already has special dispatch bytecodes that are probably about as fast as they can possibly be (short of doing type inference during compilation). Take a look here for the various kinds of optimized msg send bytecodes: https://github.com/supercollider/supercollider/blob/develop/lang/LangSource/PyrInterpreter3.cpp#L2585
It *would* be useful to add syntax to easily denote method forwarding - some kind of delegation? The only real way to do this now is via inheritance - IMO there's a over-reliance in the classlib on inheritance to share functionality between classes, so delegation could probably help there.

**Comment by telephon** [comment]
exactly, thats the point !!

**Comment by scztt** [comment]
Some made up syntax.

```
Something {
   delegatevar <>aBuffer [ Buffer:* ];
   delegatevar <>playable [ play, free ]; 
}
```

Equivalent to:
```
Something {
   play { ^playable.play }
   free { ^playable.free }
   someBufferMethod { ^aBuffer.someBufferMethod } // for every method on Buffer but not parent/sub-classes
   // arguments would need forwarding also...
}
```

Or, an alternative - allow pattern matching when defining method names. This is more flexible and has some other potentially cool uses, but might require a bit more fiddling to get optimization cases right:
```
Something {
    var aBuffer, playable;

    [play, free] {
        |...args|
        ^playable.$methodName(*args);
    }

    [Buffer:*] {
        |...args|
        ^aBuffer.$methodName(*args);
    }
}
```

**Comment by brianlheim** [comment]
> Or, an alternative - allow pattern matching when defining method names. This is more flexible and has some other potentially cool uses,

This is great!

**Comment by jamshark70** [comment]
Pattern matching is interesting, but opportunities for compiler optimization will be limited, won't they? Because you don't know until runtime which class's method table to check.

I'm not saying, don't do it, but I don't see any way that it will be fast.

Also be aware of "SendSpecialMsg" opcodes. Haven't looked into it but IIRC this is an optimization to avoid pushing commonly-used method names on the stack. If that's the only optimization, maybe it's easy but I would want to check what it's doing internally.

**Comment by telephon** [comment]
If I understand correctly, such syntax could be generally useful? 

For a first step, perhaps the delegation of methods just as a normal method call with a primitive?


```supercollider
doesNotUnderstand { |selector, args|
    this.forwardMessage(someObject, selector, args)
}

forwardMessage  { |receiver, selector, args|
    _prForwardMessage
}
```

Or are there reasons not to do this? I see that your syntax suggestion might save us one method call.

Note that not all `AbstractObject`s delegate something (there is a subclass `Delegator` in my library, which shows a possible functionality).

Maybe whoever likes could have a closer look at the  https://github.com/telephon/Neutral library? There you find already some trade-offs and perhaps you have some new ideas. It would help us to discuss this PR in detail.

**Comment by scztt** [comment]
The optimization I'm thinking of is: for any method that the compiler understands as a forward / dispatch method - things like `method { this.otherMethod }` and `method { instanceVar.method }`... there are 4 or 5 different versions it recognizes. As long as the forwarding methods we generate match these patterns (which they will), we get the optimization.

In terms of pattern matching - I'm JUST thinking of using the list of methods of a class to auto-generate forwarding methods. If you did something like this with my silly syntax example: `[Buffer:*] { ^aBuffer.$methodName }`, it wouldn't guarantee that aBuffer is a buffer or can respond to all those method names - it'll JUST look up all of Buffer's methods and generate forwarding methods to aBuffer. SC's dynamic typing means we'll never have compile-time guarantees of type, but it's still totally normal to make assumptions about the kind of objects that an instance variable may contain.

It's not so much to save a method call (it will be the same bytecode / exactly as optimal as writing it by hand). The advantage is code clarity and composability: for example, suppose you want something that behaves like an Environment, but also represents a Node on the server. Now you'd need to make your object a subclass of one or the other (which may or may not be a generally good design decision...) - and then hand-write forwarding methods for ALL of the required methods of the other class. With some kind of pattern-matching dispatch thingy, you could instead do e.g.:
```
[Node:*] { ^theNode.$methodName }
[Environment:*] { ^theEnvironment.$methodName }
```
... or subclass one but delegate the other etc. 

This would make it easy to build utility classes without needing requiring a class hierarchy.  Another example (more made-up syntax), something like:
```
Algorithms {
   *sort { |collection| ... }
   *reverse { |collection| ... }
}
CollectionIshClass {
    [MetaAlgorithms:*] { |...args| ^Algorithms.$methodName(this, *args) }
}
```
A one-line definition lets you mix in a bunch of behavior without needing to have Algorithms in your class hierarchy, and with near-zero performance penalty.

The advantages of all this: 
1. it's already highly optimized, with no additional work required
2. it would require only a small parser change, plus a corresponding compiler change.
3. the testing surface would be very low, since we'd JUST be generating an AST pattern that we already use all over the place anyway.

**Comment by telephon** [comment]
I see, you are proposing an alternative to inheritance. It is generally unrelated to the PR, it could be used anywhere. It is a little OT, but let's keep it here for now.

This would be very useful, and would make it easier to practically work with `AbstractObject`s. 

`[Meta_Algorithms:*] { |...args| ^Algorithms.$methodName(this, *args) }`

Do we need the extra syntax? There are already keywords like `thisMethod`, so wouldn't it be possible to write:

`[Meta_Algorithms:*] { |...args| ^Algorithms.performForwardedMessage(this, *args) }`

**Comment by telephon** [comment]
>In terms of pattern matching - I'm JUST thinking of using the list of methods of a class to auto-generate forwarding methods. 

am I right to think that "pattern matching" is a little bit misleading? We basically just match on a combination of a class and a selector. We are not matching on parts of a name, e.g. This makes sense: everything else would become really unwieldy very quickly.

**Comment by jamshark70** [comment]
A bit late, but:

> In terms of pattern matching - I'm JUST thinking of using the list of methods of a class to auto-generate forwarding methods.

Ah, ok, that makes sense. Now that I think about it, there's no need to forward a message that won't be understood.

Would this match methods that the target class inherits from superclasses?

Also be careful to disallow circular references -- class A imports from class B, and class B imports from A. I'm certain it can be handled gracefully, perhaps with only a non-fatal warning, but not handling it at all could easily produce infinite recursion.

> am I right to think that "pattern matching" is a little bit misleading?

I got confused thinking it would be the interpreter that matches the selector for a given occurrence of a method call against methods in the class library with wildcards. But now I understand, Scott means, the compiler searches the delegation-target class's method list for methods that match the pattern. That would require no change to the runtime at all.

**Comment by telephon** [comment]
@scztt  Am I right that the syntax is a more compact and efficient replacement of something like this?

```supercollider



Delegator : AbstractObject {
	var <>delegates, <>selectors;
	
	
	doesNotUnderstand { | selector ... args |
		var delegateClass;
		delegateClass = delegates.detect { |class| 
			var sel = selectors[class];
			(sel.isNil or: { sel.includes(selector) })
			and:
			{ class.respondsTo(selector) }
		};
		^if(delegateClass.notNil) {
			this.sendSelfMessage(delegateClass, selector, args)
		} {
			super.doesNotUnderstand(selector, *args)
		}
	}
}
```

(assuming that we had a method `sendSelfMessage` which calls a method of self from another object).

So this is basically import of multiple interfaces. What is still unclear for me is what should happen  with delegate methods that assume the existence of instance variables?

**Comment by jamshark70** [comment]
> Am I right that the syntax is a more compact and efficient replacement of something like this?

I read Scott's suggestion as literally copying the byte codes from the source class -- no delegation.

The compiler would have to verify the existence of all required instance variables, I think... too dangerous to press ahead without all variables being declared.

It's not a bad idea, but will take more creative thought about the ways users will mess it up

**Comment by jamshark70** [comment]
Sorry... once again, my phone put the "move insertion point" widget right over the "close and comment" button, making errors almost inevitable... anyway, that's all I was going to say.

**Comment by telephon** [comment]
ok, now i understand.

**Comment by telephon** [comment]
... and you want to retain the ability to override some of these in the class definition of subclasses at least.

**Comment by scztt** [comment]
We can’t copy byte codes from the “target” of the delegation, because the target depends on the type of whoever it’s being delegated to, which is only known at runtime. My suggestion is to simply expand something like:
`[method1, method2, method3] { ^delegate.$methodName } `
To
```
method1 { ^delegate.method1 }
method2 { ^delegate.method2 }
method3 { ^delegate.method3 }
```
... with whatever syntax we decide on that would make auto-delegating a whole bunch of methods both easy and maintainable. I suggested `MyClass:*` because it opens the door to mixin delegate classes for certain kinds of behavior, which would drastically reduce the need for the kinds of huge nested class hierarchys that are in the classlib now.
 Doing the above is more or less a parser change - we already compile this kind of delegation into optimized bytecode, so no compiler/bytecode changes are required.

**Comment by telephon** [comment]
Important for this PR is also: #3567.

**Comment by telephon** [comment]
Just to clarify: my suggestion of a `_forwardMessage` primitive is independent of the forwarding syntax suggestion (the syntactic extension is a very good idea nevertheless). 

I was mainly thinking of an optimization of cases where the very same selector and argument stack is called on a different object than the receiver. The receiver can call this and be sure that this is not much overhead.

It would involve a method along the lines of:
`HOT void sendSuperMessage(VMGlobals *g, PyrSymbol *selector, long numArgsPushed)`
(https://github.com/supercollider/supercollider/blob/develop/lang/LangSource/PyrMessage.cpp#L506)
a bit similar to the case:
`case methRedirect : /* send a different selector to self, e.g. this.subclassResponsibility */`
(https://github.com/supercollider/supercollider/blob/develop/lang/LangSource/PyrMessage.cpp#L588)

and

`HOT void executeMethod(VMGlobals *g, PyrMethod *meth, long numArgsPushed)`
https://github.com/supercollider/supercollider/blob/ac57a98ffbf67f6934b32cb01ca6545baa8abc73/lang/LangSource/PyrMessage.cpp#L993-L998

(with a swapped caller on the frame)

**Comment by telephon** [comment]
In any case, the current issue is mainly about the extension of the class structure with one superclass `AbstractObject` above `Object`. The superclass of `AbstractObject` is `nil`, as it is now with `Object`.

Its class `Meta_AbstractObject` is a subclass of `Class`, which is in turn an instance of `Object` (not of `AbstractObject`). `Object` remains the default inheritance of course, unless otherwise stated. So not too much is changed.

It might be discussed whether we want an `AbstractClass` object, but I'd propose not to do this. The messages that `Class` understands are already mainly for introspection, and while one might want to build an alternative system, I think it is OK to live with this trade off.

**Comment by telephon** [comment]
Here is a prepared RFC: https://github.com/telephon/rfcs/blob/master/rfcs/0002-abstract-object.md

**Comment by JordanHendersonMusic** [comment]
Just to summarise and expand on #3567 as I've been thinking about this a bit more recently as relates here as it impacts the ability to redefine certain methods...

You can't reliably overload certain methods as some are [implemented in the compiler.](https://github.com/supercollider/supercollider/blob/60b291ab18bb21c868748d8f93e5b357c435813a/lang/LangSource/PyrParseNode.cpp#L1823) 


These methods are: `if`, `case`, `switch`, `while`, `loop`, `and`, `or`, `?`, `??`, `!`, and `!?`.
This could also be the case
I'll summarise [how this works for the `if` case](https://github.com/supercollider/supercollider/blob/60b291ab18bb21c868748d8f93e5b357c435813a/lang/LangSource/PyrParseNode.cpp#L2384)  in c++ using overloads (which match the most specific signature). Each method is a little different but the idea is the same.

```c++

// 1. optimised
void compileIf(auto rec, auto InlineableBlock trueCase) { 
   ... emit bytecodes equivalent to ...
   if (rec) trueCase.()
}
 
// 1. unoptimised
void compileIf(auto rec, auto trueCase) { 
   ... call 'rec.if(trueCase)' ...
 } 

// 2. optimised
void compileIf(auto rec, auto InlineableBlock trueCase, auto InlineableBlock falseCase) { 
   ... emit bytecodes equivalent to ...
   if (rec) trueCase.() else falseCase.()
}

// 2. unoptimised
void compileIf(auto rec, auto trueCase, auto falseCase) { 
   ... call 'rec.if(trueCase, falseCase)' ...
 } 

// 3. More than 2 cases
template<class...T>  requires (  sizeof...(T) > 2 ) 
void compileIf(auto rec, T... cases) { 
   ... call 'rec.if(cases...)' ...
 } 

```
In supercollider, both overloads for `1` would look like this...
```supercollider
T {
	if { |other| ^"Correct" }
}

T().if (10) // Correct. 
T().if {10} // Error: Non Boolean in test.
```
... as `(10)` is not an inlineable block, it will actually call `T.if` going through `1. unoptimised`, but in the second case it *is* an inlineable block so emits the specific byte codes by going through `1. optimised`.

All the other methods mentioned above do something similar. There is clearly a performance reason to do this, but this may not be necessary any more? I haven't benchmarked this but performance isn't supercollider's goal so removing these optimisations might be reasonable. 

Removing these optimisations should be trivial, with at least two exceptions, which is `while` and `loop`.

`while` is implemented by turning the uninlineable arguments into inlineable arguments. From `Object.while`
```supercollider
while { arg body;
   // compiler magic: the compiler inlines the following loop
   // thus an uninlinable while can be implemented using while itself
   while({ this.value }, {
	body.value
   });
}
```

Back in the C++ example, this would look like this...
```c++

// 1. optimised
void compileWhile(auto rec, auto InlineableBlock doBlock) { 
   ... emit bytecodes equivalent to ...
   while (rec) { doBlock.() }
}
 
// 1. unoptimised
void compileWhile(auto rec, auto doBlock) { 
   compileWhile(rec, turnIntoInlineableBlock(doBlock));
 } 
```

This means you would need to express the logic of while in supercollider code (without using while). `loop` has a similar issue.

---

None of this is to suggest this isn't a good idea, but that we should really be clear with the user that they shouldn't **in any circumstances** (?) be implementing these methods because it won't work as expected. IMO, this is an issue now and should be documented as such, albeit a small issues, but with an `AbstractObject` this issue will become more apparent and we'd need to make sure the user is aware of this limitation.

**Comment by telephon** [comment]
Good you come back to this, @JordanHendersonMusic.

I do think that sclang is very much optimized towards performance, and it is pretty fast in many cases. But since times have passed, some of the trade-offs can be reconsidered.

The issue you found is subtle and a little nasty because it means that two equivalent syntaxes (functional style and message passing style) are not always equivalent.

If performance weren't impacted much, it would definitely be worth removing the optimisations, let alone for this reason. But also conceptually, of course, it would be nice to have a more consistent and extendible system.

But if it is costly, it is harder to argue.

For now, also for the abstract object quark, I can put a `shouldNotImplement` on these messages. Should we decide to make AbstractObject a superclass, this would need to be done in a different way, but I am sure one could work around it.


A side note:
As I talked to James McCartney about this, he wodered if it may not be also a good idea to remove the method table and have the method lookup be done by dictionaries. This would give us the possibility to add and modify classes at runtime. The performance impact of that may be much higher of course.

**Comment by JordanHendersonMusic** [comment]
> it means that two equivalent syntaxes (functional style and message passing style) are not always equivalent.

I don't think that is quite true as it is the ininlineability of the arguments rather than the syntax that changes the behaviour. 

It is true in the case of `if`, but not `while`.

```supercollider
// message style
T().if({10}) // Error
T().if( 10 ) // Corrent
T().if{ 10 } // Error

// inside-out message style
if(T(),   10  ) // Correct
if(T(), { 10 }) // Error

// imperative style
if (T()) {  10  } // Error
if (T()) 10; // Invalid syntax - but would be correct  if possible


T().while {10.postln} // Correct - message style
while(T(), {10.postln}) // Correct - inside-out message style

{T()}.while {10.postln} // Error non boolean - message style

while {T()} {10.postln} // Error non boolean - imperative style 
while (T()) {10.postln} // Correct
while T() {10.postln} // Invalid syntax, imperative style, but would be correct if possible
```

I think there are two reasons this is confusing. 

First, supercollider doesn't have keywords and imperative style is a trick in the parser where trailing blocks are appended to the message call — it is not possible to append trailing expressions due to supercollider's design.  It is a little bit surprising that literal function objects change the parsing. This leads to the expectation that the code in curlys is a block of code (as it is in other languages), when actually it is a function object.

Here is an example.
```supercollider
T { 
   *getF { ^{ "hello" } }
   foo { |f| ^f.() ++ "world" }
}
T().foo { "hello" } == "helloworld";

T().foo T.getF // syntax error, but semantically identical.
```

Second, we have traditionally used curlys for the first argument of `while` rather than normal parenthesis. 
```
while (T()) {10.postln} // Correct
while {T()} {10.postln} // Correct
```
This is because it would require the compiler knows about the 'while' message (i.e., a keyword) and emit special instructions to re-evaluate the first argument.

**Comment by telephon** [comment]
Yes, this is a good overview – thank you! As being used to (almost) everything being a first-class object, I am mainly confused when it isn't (quite).

**Comment by JordanHendersonMusic** [comment]
I'm just collecting all the abstract object issues into one place #7217 
Lets try and keep this issue just for the creation of abstract object and discuss forwarding optimisation in separate issues, perhaps here #7216, or a new one if the optimisation is different.
